receiver/kafkareceiver/consumer_franz.go

```diff
@@ -0,0 +1,476 @@
+		// Shutdown will deadlock when `franzConsumer` inevitably calls the
+		// lost/assigned callback.
+		c.client.Close()
+	}
+	select {
+	case <-c.started:
+	default: // Return immediately if the receiver hasn't started.
+		return nil
+	}
+
+	select {
+	case <-ctx.Done():
+		c.cancelProcessing(fmt.Errorf(
+			"kafka consumer: shutdown context done: %w", ctx.Err(),
+		))
+		return ctx.Err()
+	case <-c.consumerClosed:
+	}
+	return nil
+}
+
+// assigned must be set as a kgo.OnPartitionsAssigned callback. Ensuring all
+// assigned partitions to this consumer process received records.
+func (c *franzConsumer) assigned(ctx context.Context, _ *kgo.Client, assigned map[string][]int32) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for topic, partitions := range assigned {
+		for _, partition := range partitions {
+			c.telemetryBuilder.KafkaReceiverPartitionStart.Add(ctx,
+				1, metric.WithAttributes(attribute.String(attrInstanceName, c.id.Name())),
+			)
+			c.assignments[topicPartition{topic: topic, partition: partition}] = &pc{
+				backOff: newExponentialBackOff(c.config.ErrorBackOff),
+				logger: c.logger.With(
+					zap.String(attrTopic, topic),
+					zap.String(attrPartition, strconv.Itoa(int(partition))),
+				),
+				attrs: attribute.NewSet(
+					attribute.String(attrInstanceName, c.id.String()),
+					attribute.String(attrTopic, topic),
+					attribute.String(attrPartition, strconv.Itoa(int(partition))),
+				),
+			}
+		}
+	}
+}
+
+// lost must be set as a kgo.OnPartitionsLost and kgo.OnPartitionsReassigned
+// callbacks. Ensures that partitions that are lost (see kgo.OnPartitionsLost
+// for more details) or reassigned (see kgo.OnPartitionsReassigned for more
+// details) have their partition consumer stopped.
+// This callback must finish within the re-balance timeout.
+func (c *franzConsumer) lost(_ context.Context, _ *kgo.Client, lost map[string][]int32) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for topic, partitions := range lost {
+		for _, partition := range partitions {
+			tp := topicPartition{topic: topic, partition: partition}
+			c.telemetryBuilder.KafkaReceiverPartitionClose.Add(context.Background(),
+				1, metric.WithAttributes(attribute.String(attrInstanceName, c.id.Name())),
+			)
+			delete(c.assignments, tp)
+		}
+	}
+}
+
+// handleMessage is called on a per-partition basis.
+func (c *franzConsumer) handleMessage(ctx context.Context, msg kafkaMessage, pc *pc) error {
+	defer pc.resetBackoff()
+
+	for {
+		err := c.consumeMessage(ctx, msg, pc.attrs)
+		if err == nil {
+			return nil // Successfully processed.
+		}
+		// In the future, with Consumer Share Groups, messages not processed
+		// within a configurable timeout, are re-delivered to the consumer in
+		// the Share group, however, at the time of writing this feature isn't
+		// yet GA nor widely deployed.
+		// Share groups are more in line with observability and OTel collector
+		// use cases than traditional consumer groups.
```

Comment:
The Kafka receiver may be used in analytics pipelines, so I'm not sure we could say that for all possible use cases. Perhaps worth stating that we should make it an option in the future though.